cpu/qn908x/periph/i2c.c

```diff
@@ -0,0 +1,357 @@
+     * 0xffff.
+     * The MSTTIME register controls for how many cycles does the clock stay
+     * low and high, allowing to use different values for each one but we only
+     * use symmetric ones here, which is why there's a * 2 in the formula above.
+     */
+    assert(speed_hz > 0);
+    uint32_t bus_freq = CLOCK_GetFreq(kCLOCK_BusClk);
+    uint32_t target = bus_freq / (2 * speed_hz);
+
+    uint32_t best_error = UINT_MAX;
+    uint32_t best_clkdiv = 0;
+    uint32_t best_msttime = 0;
+
+    for (uint32_t msttime_p2 = 9; msttime_p2 >= 2; msttime_p2--) {
+        uint32_t clkdiv_p1 = (target + msttime_p2 / 2) / msttime_p2;
+        if (clkdiv_p1 >= I2C_CLKDIV_MAX + 1) {
+            clkdiv_p1 = I2C_CLKDIV_MAX + 1;
+        }
+        uint32_t error =
+            abs((int32_t)target - (int32_t)(clkdiv_p1 * msttime_p2));
+        if (error < best_error) {
+            best_error = error;
+            best_clkdiv = clkdiv_p1 - 1;
+            best_msttime = msttime_p2 - 2;
+        }
+    }
+    i2c_dev->CLKDIV = I2C_CLKDIV_DIVVAL(best_clkdiv);
+    i2c_dev->MSTTIME = I2C_MSTTIME_MSTSCLLOW(best_msttime) |
+                       I2C_MSTTIME_MSTSCLHIGH(best_msttime);
+    DEBUG("[i2c]: bus_fq=%" PRIu32" target_freq=%" PRIu32" msttime=%" PRIu32
+          " clkdiv=%" PRIu32 " error=%" PRIu32 "\n",
+          bus_freq, speed_hz, best_msttime, best_clkdiv, best_error);
+}
+
+static void _i2c_init_pins(i2c_t dev)
+{
+    const i2c_conf_t *conf = &i2c_config[dev];
+
+    /* Configure SDA and SCL pins, the function value depends on the pin:
+     * FUNC4: A6, A7, A8, A9, A12, A13, A22, A23, A26, A27
+     * FUNC5: A2, A3, A4, A5, A20, A21
+     */
+    static const uint32_t func5_mask =
+        (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5) | (1 << 20) | (1 << 21);
+    /* TODO: Have a way to configure IOCON_MODE_PULLUP and IOCON_DRIVE_HIGH
+     * from the board. */
+    gpio_init_mux(conf->pin_sda,
+                  ((1u << GPIO_T_PIN(conf->pin_sda)) & func5_mask) ? 5 : 4);
+    gpio_init_mux(conf->pin_scl,
+                  ((1u << GPIO_T_PIN(conf->pin_scl)) & func5_mask) ? 5 : 4);
+
+    mutex_unlock(&locks[dev]);
+}
+
+void i2c_init(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+    const i2c_conf_t *conf = &i2c_config[dev];
+    I2C_Type *const i2c_dev = conf->dev;
+
+    int flexcomm_num = flexcomm_init((FLEXCOMM_Type *)i2c_dev, FLEXCOMM_ID_I2C);
+    assert(flexcomm_num >= 0);
+
+    if (flexcomm_num == 1) {
+        /* Disable the FLECOMM1 MULT1/DIV1 divisor. FLEXCOMM2 doesn't have a
+         * fractional divisor. This divides the clock by (1 + N / 256) where
+         * N is set to 0 here, so the fractional divisor only divides by one. */
+        CLOCK_SetClkDiv(kCLOCK_DivFrg1, 0u);
+    }
+
+    /* Enable controller mode, no timeout, no monitor, no clock stretching. */
+    i2c_dev->CFG = I2C_CFG_MSTEN_MASK;
+    _i2c_controller_set_speed(i2c_dev, conf->speed * 1000);
```

Comment:
Makes me wonder: why not have `i2c_speed_t` as bit/s 
That would be consistent with other architectures that allow just any value here.

Reply:
Sure.